control-plane/connect-inject/common/common_test.go

```diff
@@ -399,3 +407,104 @@ func TestGetPortProtocol(t *testing.T) {
+		input                  error
+		expectMissingNamespace bool
+	}{
+		{
+			name:                   "nil error",
+			expectMissingNamespace: false,
+		},
+		{
+			name:                   "random error",
+			input:                  fmt.Errorf("namespace resource not found"),
+			expectMissingNamespace: false,
+		},
+		{
+			name:                   "grpc code is not InvalidArgument",
+			input:                  status.Error(codes.NotFound, "namespace resource not found"),
+			expectMissingNamespace: false,
+		},
+		{
+			name:                   "grpc code is InvalidArgument, but the message is not for namespaces",
+			input:                  status.Error(codes.InvalidArgument, "blurg resource not found"),
+			expectMissingNamespace: false,
+		},
+		{
+			name:                   "namespace is missing",
+			input:                  status.Error(codes.InvalidArgument, "namespace resource not found"),
+			expectMissingNamespace: true,
+		},
+	}
+
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			actual := ConsulNamespaceIsNotFound(tt.input)
+			require.Equal(t, tt.expectMissingNamespace, actual)
+		})
+	}
+}
+
+// Test_ConsulNamespaceIsNotFound_ErrorMsg is an integration test that verifies the error message
+// associated with a missing namespace while creating a resource doesn't drift.
+func Test_ConsulNamespaceIsNotFound_ErrorMsg(t *testing.T) {
+	t.Parallel()
+
+	// Create test consulServer server.
+	testClient := test.TestServerWithMockConnMgrWatcher(t, func(c *testutil.TestServerConfig) {
+		c.Experiments = []string{"resource-apis"}
+	})
+	resourceClient, err := consul.NewResourceServiceClient(testClient.Watcher)
+	require.NoError(t, err)
+
+	id := &pbresource.ID{
+		Name: "foo",
+		Type: &pbresource.Type{
+			Group:        "catalog",
+			GroupVersion: "v1alpha1",
+			Kind:         "Workload",
+		},
+		Tenancy: &pbresource.Tenancy{
+			Partition: constants.DefaultConsulPartition,
+			Namespace: "i-dont-exist-but-its-ok-we-will-meet-again-someday",
+
+			// Because we are explicitly defining NS/partition, this will not default and must be explicit.
+			// At a future point, this will move out of the Tenancy block.
+			PeerName: constants.DefaultConsulPeer,
+		},
+	}
+
+	workload := &pbcatalog.Workload{
+		Addresses: []*pbcatalog.WorkloadAddress{
+			{Host: "10.0.0.1", Ports: []string{"mesh"}},
+		},
+		Ports: map[string]*pbcatalog.WorkloadPort{
+			"mesh": {
+				Port:     constants.ProxyDefaultInboundPort,
+				Protocol: pbcatalog.Protocol_PROTOCOL_MESH,
+			},
+		},
+		NodeName: "banana",
+		Identity: "foo",
+	}
+
+	data := ToProtoAny(workload)
+
+	resource := &pbresource.Resource{
+		Id:   id,
+		Data: data,
+	}
+
+	_, err = resourceClient.Write(context.Background(), &pbresource.WriteRequest{Resource: resource})
+	require.Error(t, err)
+
+	s, ok := status.FromError(err)
+	require.True(t, ok)
+	require.Equal(t, codes.InvalidArgument, s.Code())
+	require.Contains(t, s.Message(), "namespace resource not found")
```

Comment:
Love this test! Only suggestion I have is to also assert `ConsulNamespaceIsNotFound` returns true here